<!DOCTYPE HTML>
<html lang="en">
<head>
<!-- Generated by javadoc (17) -->
<title>io.jenetics.prog (Jenetics 7.2.0)</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="description" content="declaration: module: io.jenetics.prog, package: io.jenetics.prog">
<meta name="generator" content="javadoc/PackageWriterImpl">
<link rel="stylesheet" type="text/css" href="../../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../../../../script.js"></script>
<script type="text/javascript" src="../../../../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../../../../script-dir/jquery-ui.min.js"></script>
</head>
<body class="package-declaration-page">
<script type="text/javascript">var pathtoroot = "../../../../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top">
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../../../../index.html">Overview</a></li>
<li><a href="../../../module-summary.html">Module</a></li>
<li class="nav-bar-cell1-rev">Package</li>
<li>Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../index-all.html">Index</a></li>
<li><a href="../../../../help-doc.html#package">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Package:&nbsp;</li>
<li><a href="#package-description">Description</a>&nbsp;|&nbsp;</li>
<li><a href="#related-package-summary">Related Packages</a>&nbsp;|&nbsp;</li>
<li><a href="#class-summary">Classes and Interfaces</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" value="search" disabled="disabled">
<input type="reset" id="reset-button" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<div class="header">
<div class="sub-title"><span class="module-label-in-package">Module</span>&nbsp;<a href="../../../module-summary.html">io.jenetics.prog</a></div>
<h1 title="Package io.jenetics.prog" class="title">Package io.jenetics.prog</h1>
</div>
<hr>
<div class="package-signature">package <span class="element-name">io.jenetics.prog</span></div>
<section class="package-description" id="package-description">
<div class="block"><h2>Example</h2>

 The following example shows how to solve a GP problem with <em>Jenetics</em>.
 We are trying to find a polynomial (or an arbitrary mathematical function)
 which approximates a given data set.

 <table>
     <caption>Sample points</caption>
     <tr><th>x</th><th>y</th></tr>
     <tr><td>0.00</td><td>0.0000</td></tr>
     <tr><td>0.10</td><td>0.0740</td></tr>
     <tr><td>0.20</td><td>0.1120</td></tr>
     <tr><td>0.30</td><td>0.1380</td></tr>
     <tr><td>...</td><td>...</td></tr>
 </table>

 The sample points has been created with the function
 <em>f(x) = 4*x^3 - 3*x^2 + x</em>. The knowledge of the creating function
 makes it easier to compare the quality of the evolved function. For the
 example we created 21 data points.

 <p>
     <b>NOTE</b>: <em>The function which created the sample points is not
     needed in the error function we have to define for the GP. It just let
     us verify the final, evolved result.</em>
 </p>

 As first step, we have to define the set of allowed <em>non-terminal</em>
 and the <em>terminal</em> operations the GP is working with. Selecting the
 right set of operation has a big influence on the performance of the GP. If
 the operation set is bigger than necessary, we will expand the potential
 search space, and the execution time for finding a solution. For our
 <em>polynomial</em> example we will chose the following <em>operations</em>
 and <em>terminals</em>.

 <div class="code"><code lang="java"> <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>final</b></font> ISeq&lt;Op&lt;Double&gt;&gt; OPERATIONS = ISeq.of(
     MathOp.ADD,
     MathOp.SUB,
     MathOp.MUL
 );

 <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>final</b></font> ISeq&lt;Op&lt;Double&gt;&gt; TERMINALS = ISeq.of(
     Var.of(<font color="#0000FF">"x"</font>, 0),
     EphemeralConst.of(() -&gt; (<font color="#7F0055"><b>double</b></font>)RandomRegistry.getRandom().nextInt(10))
 );</code></div>

 The chosen non-terminal operation set is sufficient to create any polynomial.
 For the terminal operations, we added a variable "x", with index zero, and
 an ephemeral integer constant. The purpose of the ephemeral constant is to
 create constant values, which will differ for every tree, but stay constant
 within a tree.
 <p>
 In the next step define the fitness function for the GP, which will be an
 error function we will minimize.

 <div class="code"><code lang="java"> <font color="#3F7F5F">// The lookup table where the data points are stored.</font>
 <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>final</b></font> <font color="#7F0055"><b>double</b></font>[][] SAMPLES = <font color="#7F0055"><b>new</b></font> <font color="#7F0055"><b>double</b></font>[][] {
     {-1.0, -8.0000},
     {-0.9, -6.2460},
     ...
 };

 <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>double</b></font> error(<font color="#7F0055"><b>final</b></font> ProgramGene&lt;Double&gt; program) {
     <font color="#7F0055"><b>return</b></font> Arrays.stream(SAMPLES).mapToDouble(sample -&gt; {
         <font color="#7F0055"><b>final</b></font> <font color="#7F0055"><b>double</b></font> x = sample[0];
         <font color="#7F0055"><b>final</b></font> <font color="#7F0055"><b>double</b></font> y = sample[1];
         <font color="#7F0055"><b>final</b></font> <font color="#7F0055"><b>double</b></font> result = program.eval(x);
         <font color="#7F0055"><b>return</b></font> abs(y - result) + program.size()*0.00005;
     })
     .sum();
 }</code></div>

 The error function calculates the sum of the (absolute) difference between
 the sample value and the value calculated the by the evolved <em>program</em>
 (<code>ProgramGene</code>). Since we prefer compact programs over complex one, we
 will add a penalty for the program size (the number of nodes of the program
 tree).

 <p>
     <b>CAUTION</b>: <em>The penalty for the tree size must be small enough to
     not dominate the error function. We still want to find an approximating
     function and not the smallest possible one</em>
 </p>

 After we have defined the error function, we need to define the proper
 <code>Codec</code>.

 <div class="code"><code lang="java"> <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>final</b></font> Codec&lt;ProgramGene&lt;Double&gt;, ProgramGene&lt;Double&gt;&gt; CODEC =
     Codec.of(
         Genotype.of(ProgramChromosome.of(
             <font color="#3F7F5F">// Program tree depth.</font>
             5,
             <font color="#3F7F5F">// Chromosome validator.</font>
             ch -&gt; ch.root().size() &lt;= 50,
             OPERATIONS,
             TERMINALS
         )),
         Genotype::gene
     );</code></div>


 There are two particularities in the definition of the
 <code>ProgramChromosome</code>:

 <ol>
     <li>Since we want to narrow the search space, we are limiting the depth
     of newly created program trees to 5.</li>
     <li>Because of crossover operations performed during evolution, the
     resulting programs can grow quite big. To prevent an unlimited growth of
     the program trees, we mark programs with more than _50_ nodes as
     invalid.</li>
 </ol>

 Now we are ready to put everything together:

 <div class="code"><code lang="java"> <font color="#7F0055"><b>public</b></font> <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>void</b></font> main(<font color="#7F0055"><b>final</b></font> String[] args) {
     <font color="#7F0055"><b>final</b></font> Engine&lt;ProgramGene&lt;Double&gt;, Double&gt; engine = Engine
         .builder(Polynomial::error, CODEC)
         .minimizing()
         .alterers(
             <font color="#7F0055"><b>new</b></font> SingleNodeCrossover&lt;&gt;(),
             <font color="#7F0055"><b>new</b></font> Mutator&lt;&gt;())
         .build();

     <font color="#7F0055"><b>final</b></font> ProgramGene&lt;Double&gt; program = engine.stream()
         .limit(500)
         .collect(EvolutionResult.toBestGenotype())
         .gene();

     System.out.println(Tree.toString(program));
 }</code></div>

 The GP is capable of finding the polynomial which created the sample data.
 After a few tries, we got the following (correct) output program:

 <pre>
 add
 ├── mul
 │   ├── x
 │   └── sub
 │       ├── 0.0
 │       └── mul
 │           ├── x
 │           └── sub
 │               ├── sub
 │               │   ├── sub
 │               │   │   ├── sub
 │               │   │   │   ├── 3.0
 │               │   │   │   └── x
 │               │   │   └── x
 │               │   └── x
 │               └── x
 └── x
 </pre>

 This program can be reduced to <em>4*x^3 - 3*x^2 + x</em>, which is exactly
 the polynomial, which created the sample data.</div>
<dl class="notes">
<dt>Since:</dt>
<dd>3.9</dd>
<dt>Version:</dt>
<dd>3.9</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<li>
<div id="related-package-summary">
<div class="caption"><span>Related Packages</span></div>
<div class="summary-table three-column-summary">
<div class="table-header col-first">Module</div>
<div class="table-header col-second">Package</div>
<div class="table-header col-last">Description</div>
<div class="col-plain even-row-color"><a href="../../../../io.jenetics.base/module-summary.html">io.jenetics.base</a></div>
<div class="col-first even-row-color"><a href="../../../../io.jenetics.base/io/jenetics/package-summary.html">io.jenetics</a></div>
<div class="col-last even-row-color">
<div class="block">This is the base package of the Jenetics library and contains all domain
 classes, like Gene, Chromosome or Genotype.</div>
</div>
<div class="col-plain odd-row-color"><a href="../../../module-summary.html">io.jenetics.prog</a></div>
<div class="col-first odd-row-color"><a href="op/package-summary.html">io.jenetics.prog.op</a></div>
<div class="col-last odd-row-color">
<div class="block">Operations</div>
</div>
<div class="col-plain even-row-color"><a href="../../../module-summary.html">io.jenetics.prog</a></div>
<div class="col-first even-row-color"><a href="regression/package-summary.html">io.jenetics.prog.regression</a></div>
<div class="col-last even-row-color">&nbsp;</div>
<div class="col-plain odd-row-color"><a href="../../../../io.jenetics.base/module-summary.html">io.jenetics.base</a></div>
<div class="col-first odd-row-color"><a href="../../../../io.jenetics.base/io/jenetics/engine/package-summary.html">io.jenetics.engine</a></div>
<div class="col-last odd-row-color">
<div class="block">This package contains classes which perform the actual evolution.</div>
</div>
<div class="col-plain even-row-color"><a href="../../../../io.jenetics.ext/module-summary.html">io.jenetics.ext</a></div>
<div class="col-first even-row-color"><a href="../../../../io.jenetics.ext/io/jenetics/ext/package-summary.html">io.jenetics.ext</a></div>
<div class="col-last even-row-color">
<div class="block">This package contains additional genes/chromosomes and GA operators.</div>
</div>
<div class="col-plain odd-row-color"><a href="../../../../io.jenetics.base/module-summary.html">io.jenetics.base</a></div>
<div class="col-first odd-row-color"><a href="../../../../io.jenetics.base/io/jenetics/stat/package-summary.html">io.jenetics.stat</a></div>
<div class="col-last odd-row-color">
<div class="block">This package contains additional statistics classes which are not available
 in the Java core library.</div>
</div>
<div class="col-plain even-row-color"><a href="../../../../io.jenetics.base/module-summary.html">io.jenetics.base</a></div>
<div class="col-first even-row-color"><a href="../../../../io.jenetics.base/io/jenetics/util/package-summary.html">io.jenetics.util</a></div>
<div class="col-last even-row-color">
<div class="block">This package contains the collection classes (<a href="../../../../io.jenetics.base/io/jenetics/util/Seq.html" title="interface in io.jenetics.util"><code>Seq</code></a>,
 <a href="../../../../io.jenetics.base/io/jenetics/util/ISeq.html" title="interface in io.jenetics.util"><code>ISeq</code></a> and <a href="../../../../io.jenetics.base/io/jenetics/util/MSeq.html" title="interface in io.jenetics.util"><code>MSeq</code></a>) which are
 used in the public interfaces of the <a href="../../../../io.jenetics.base/io/jenetics/Chromosome.html" title="interface in io.jenetics"><code>Chromosome</code></a> and
 <a href="../../../../io.jenetics.base/io/jenetics/Genotype.html" title="class in io.jenetics"><code>Genotype</code></a>.</div>
</div>
<div class="col-plain odd-row-color"><a href="../../../../io.jenetics.xml/module-summary.html">io.jenetics.xml</a></div>
<div class="col-first odd-row-color"><a href="../../../../io.jenetics.xml/io/jenetics/xml/package-summary.html">io.jenetics.xml</a></div>
<div class="col-last odd-row-color">&nbsp;</div>
</div>
</div>
</li>
<li>
<div id="class-summary">
<div class="caption"><span>Classes</span></div>
<div class="summary-table two-column-summary">
<div class="table-header col-first">Class</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="MathRewriteAlterer.html" title="class in io.jenetics.prog">MathRewriteAlterer</a>&lt;G extends <a href="../../../../io.jenetics.ext/io/jenetics/ext/TreeGene.html" title="interface in io.jenetics.ext">TreeGene</a>&lt;<a href="op/Op.html" title="interface in io.jenetics.prog.op">Op</a>&lt;<a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html" title="class or interface in java.lang" class="external-link">Double</a>&gt;,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Prunes a given mathematical tree with the given alterer probability.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="ProgramChromosome.html" title="class in io.jenetics.prog">ProgramChromosome</a>&lt;A&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">Holds the nodes of the operation tree.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="ProgramGene.html" title="class in io.jenetics.prog">ProgramGene</a>&lt;A&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">This gene represents a program, build upon an AST of <a href="op/Op.html" title="interface in io.jenetics.prog.op"><code>Op</code></a> functions.</div>
</div>
</div>
</div>
</li>
</ul>
</section>
</main>
<footer role="contentinfo">
<hr>
<p class="legal-copy"><small>&copy; 2007-2023 Franz Wilhelmst&ouml;tter  &nbsp;<i>(2023-08-30 19:51)</i></small></p>
</footer>
</div>
</div>
</body>
</html>
