<!DOCTYPE HTML>
<html lang="en">
<head>
<!-- Generated by javadoc (17) -->
<title>io.jenetics (Jenetics 7.0.0)</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="description" content="declaration: module: io.jenetics.base, package: io.jenetics">
<meta name="generator" content="javadoc/PackageWriterImpl">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../script-dir/jquery-ui.min.css" title="Style">
<link rel="stylesheet" type="text/css" href="../../../jquery-ui.overrides.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-3.5.1.min.js"></script>
<script type="text/javascript" src="../../../script-dir/jquery-ui.min.js"></script>
</head>
<body class="package-declaration-page">
<script type="text/javascript">var evenRowColor = "even-row-color";
var oddRowColor = "odd-row-color";
var tableTab = "table-tab";
var activeTableTab = "active-table-tab";
var pathtoroot = "../../../";
loadScripts(document, 'script');</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<div class="flex-box">
<header role="banner" class="flex-header">
<nav role="navigation">
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="top-nav" id="navbar-top">
<div class="skip-nav"><a href="#skip-navbar-top" title="Skip navigation links">Skip navigation links</a></div>
<ul id="navbar-top-firstrow" class="nav-list" title="Navigation">
<li><a href="../../module-summary.html">Module</a></li>
<li class="nav-bar-cell1-rev">Package</li>
<li>Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../index-all.html">Index</a></li>
<li><a href="../../../help-doc.html#package">Help</a></li>
</ul>
</div>
<div class="sub-nav">
<div>
<ul class="sub-nav-list">
<li>Package:&nbsp;</li>
<li><a href="#package-description">Description</a>&nbsp;|&nbsp;</li>
<li><a href="#related-package-summary">Related Packages</a>&nbsp;|&nbsp;</li>
<li><a href="#class-summary">Classes and Interfaces</a></li>
</ul>
</div>
<div class="nav-list-search"><label for="search-input">SEARCH:</label>
<input type="text" id="search-input" value="search" disabled="disabled">
<input type="reset" id="reset-button" value="reset" disabled="disabled">
</div>
</div>
<!-- ========= END OF TOP NAVBAR ========= -->
<span class="skip-nav" id="skip-navbar-top"></span></nav>
</header>
<div class="flex-content">
<main role="main">
<div class="header">
<div class="sub-title"><span class="module-label-in-package">Module</span>&nbsp;<a href="../../module-summary.html">io.jenetics.base</a></div>
<h1 title="Package io.jenetics" class="title">Package io.jenetics</h1>
</div>
<hr>
<div class="package-signature">package <span class="element-name">io.jenetics</span></div>
<section class="package-description" id="package-description">
<div class="block">This is the base package of the Jenetics library and contains all domain
 classes, like Gene, Chromosome or Genotype. Most of this types are immutable
 data classes and doesn't implement any behavior. It also contains the Selector
 and Alterer interfaces and its implementations. The classes in this package
 are (almost) sufficient to implement an own GA.

 <b>Introduction</b>
 <p><strong>Jenetics</strong> is an <strong>Genetic Algorithm</strong>,
 respectively an <strong>Evolutionary Algorithm</strong>, library written in
 Java. It is designed with a clear separation of the several concepts of the
 algorithm, e.&nbsp;g. Gene, Chromosome, Genotype, Phenotype, Population and
 fitness Function. <strong>Jenetics</strong> allows you to minimize and
 maximize the given fitness function without tweaking it. In contrast to other
 GA implementations, the library uses the concept of an evolution stream
 (EvolutionStream) for executing the evolution steps. Since the
 EvolutionStream implements the Java Stream interface, it works smoothly with
 the rest of the Java streaming API.</p>

 <b>Getting Started</b>
 <p>The minimum evolution Engine setup needs a genotype factory,
 Factory&lt;Genotype&lt;?&gt;&gt;, and a fitness Function. The Genotype
 implements the Factory interface and can therefore be used as prototype for
 creating the initial Population and for creating new random Genotypes.</p>

 <div class="code"><code lang="java"> <font color="#7F0055"><b>import</b></font> io.jenetics.BitChromosome;
 <font color="#7F0055"><b>import</b></font> io.jenetics.BitGene;
 <font color="#7F0055"><b>import</b></font> io.jenetics.Genotype;
 <font color="#7F0055"><b>import</b></font> io.jenetics.engine.Engine;
 <font color="#7F0055"><b>import</b></font> io.jenetics.engine.EvolutionResult;
 <font color="#7F0055"><b>import</b></font> io.jenetics.util.Factory;

 <font color="#7F0055"><b>public</b></font> <font color="#7F0055"><b>class</b></font> HelloWorld {
      <font color="#3F7F5F">// 2.) Definition of the fitness function.</font>
     <font color="#7F0055"><b>private</b></font> <font color="#7F0055"><b>static</b></font> Integer eval(Genotype&lt;BitGene&gt; gt) {
         <font color="#7F0055"><b>return</b></font> ((BitChromosome)gt.chromosome()).bitCount();
     }

     <font color="#7F0055"><b>public</b></font> <font color="#7F0055"><b>static</b></font> <font color="#7F0055"><b>void</b></font> main(String[] args) {
         <font color="#3F7F5F">// 1.) Define the genotype (factory) suitable</font>
         <font color="#3F7F5F">//     for the problem.</font>
         Factory&lt;Genotype&lt;BitGene&gt;&gt; gtf =
         Genotype.of(BitChromosome.of(10, 0.5));

         <font color="#3F7F5F">// 3.) Create the execution environment.</font>
         Engine&lt;BitGene, Integer&gt; engine = Engine
            .builder(HelloWorld::eval, gtf)
            .build();

         <font color="#3F7F5F">// 4.) Start the execution (evolution) and</font>
         <font color="#3F7F5F">//     collect the result.</font>
         Genotype&lt;BitGene&gt; result = engine.stream()
             .limit(100)
             .collect(EvolutionResult.toBestGenotype());

         System.out.println(<font color="#0000FF">"Hello World:\n"</font> + result);
     }
 }</code></div>

 <p>In contrast to other GA implementations, the library uses the concept of
 an evolution stream (EvolutionStream) for executing the evolution steps.
 Since the EvolutionStream implements the Java Stream interface, it works
 smoothly with the rest of the Java streaming API. Now let's have a closer
 look at listing above and discuss this simple program step by step:</p>
 <ol>
 <li>The probably most challenging part, when setting up a new evolution
 Engine, is to transform the problem domain into a appropriate Genotype
 (factory) representation. In our example we want to count the number of ones
 of a BitChromosome. Since we are counting only the ones of one chromosome,
 we are adding only one BitChromosome to our Genotype. In general, the
 Genotype can be created with 1 to n chromosomes.</li>
 <li>Once this is done, the fitness function which should be maximized, can be
 defined. Utilizing the new language features introduced in Java 8, we simply
 write a private static method, which takes the genotype we defined and
 calculate it's fitness value. If we want to use the optimized bit-counting
 method, bitCount(), we have to cast the Chromosome&lt;BitGene&gt; class to
 the actual used BitChromosome class. Since we know for sure that we created
 the Genotype with a BitChromosome, this can be done safely. A reference to
 the eval method is then used as fitness function and passed to the
 Engine.build method.</li>
 <li>In the third step we are creating the evolution Engine, which is
 responsible for changing, respectively evolving, a given population. The
 Engine is highly configurable and takes parameters for controlling the
 evolutionary and the computational environment. For changing the evolutionary
 behavior, you can set different alterers and selectors. By changing the used
 Executor service, you control the number of threads, the Engine is allowed to
 use. An new Engine instance can only be created via its builder, which is
 created by calling the Engine.builder method.</li>
 <li>In the last step, we can create a new EvolutionStream from our Engine.
 The EvolutionStream is the model or view of the evolutionary process. It
 serves as a »process handle« and also allows you, among other things, to
 control the termination of the evolution. In our example, we simply truncate
 the stream after 100 generations. If you don't limit the stream, the
 EvolutionStream will not terminate and run forever. Since the EvolutionStream
 extends the java.util.stream.Stream interface, it integrates smoothly with
 the rest of the Java streaming API. The final result, the best Genotype in
 our example, is then collected with one of the predefined collectors of the
 EvolutionResult class.</li>
 </ol></div>
<dl class="notes">
<dt>Since:</dt>
<dd>1.0</dd>
<dt>Version:</dt>
<dd>3.1</dd>
</dl>
</section>
<section class="summary">
<ul class="summary-list">
<li>
<div id="related-package-summary">
<div class="caption"><span>Related Packages</span></div>
<div class="summary-table two-column-summary">
<div class="table-header col-first">Package</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color"><a href="engine/package-summary.html">io.jenetics.engine</a></div>
<div class="col-last even-row-color">
<div class="block">This package contains classes which performs the actual evolution.</div>
</div>
<div class="col-first odd-row-color"><a href="stat/package-summary.html">io.jenetics.stat</a></div>
<div class="col-last odd-row-color">
<div class="block">This package contains additional statistics classes which are not available
 in the Java core library.</div>
</div>
<div class="col-first even-row-color"><a href="util/package-summary.html">io.jenetics.util</a></div>
<div class="col-last even-row-color">
<div class="block">This package contains the collection classes (<a href="util/Seq.html" title="interface in io.jenetics.util"><code>Seq</code></a>,
 <a href="util/ISeq.html" title="interface in io.jenetics.util"><code>ISeq</code></a> and <a href="util/MSeq.html" title="interface in io.jenetics.util"><code>MSeq</code></a>) which are
 used in the public interfaces of the <a href="Chromosome.html" title="interface in io.jenetics"><code>Chromosome</code></a> and
 <a href="Genotype.html" title="class in io.jenetics"><code>Genotype</code></a>.</div>
</div>
</div>
</div>
</li>
<li>
<div id="class-summary">
<div class="table-tabs" role="tablist" aria-orientation="horizontal"><button id="class-summary-tab0" role="tab" aria-selected="true" aria-controls="class-summary.tabpanel" tabindex="0" onkeydown="switchTab(event)" onclick="show('class-summary', 'class-summary', 2)" class="active-table-tab">All Classes and Interfaces</button><button id="class-summary-tab1" role="tab" aria-selected="false" aria-controls="class-summary.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('class-summary', 'class-summary-tab1', 2)" class="table-tab">Interfaces</button><button id="class-summary-tab2" role="tab" aria-selected="false" aria-controls="class-summary.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('class-summary', 'class-summary-tab2', 2)" class="table-tab">Classes</button><button id="class-summary-tab3" role="tab" aria-selected="false" aria-controls="class-summary.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('class-summary', 'class-summary-tab3', 2)" class="table-tab">Enum Classes</button><button id="class-summary-tab4" role="tab" aria-selected="false" aria-controls="class-summary.tabpanel" tabindex="-1" onkeydown="switchTab(event)" onclick="show('class-summary', 'class-summary-tab4', 2)" class="table-tab">Record Classes</button></div>
<div id="class-summary.tabpanel" role="tabpanel">
<div class="summary-table two-column-summary" aria-labelledby="class-summary-tab0">
<div class="table-header col-first">Class</div>
<div class="table-header col-last">Description</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="AbstractAlterer.html" title="class in io.jenetics">AbstractAlterer</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Abstract implementation of the alterer interface.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="AbstractChromosome.html" title="class in io.jenetics">AbstractChromosome</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">The abstract base implementation of the Chromosome interface.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab1"><a href="Alterer.html" title="interface in io.jenetics">Alterer</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab1">
<div class="block">The Alterer is responsible for the changing/recombining the Population.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab4"><a href="AltererResult.html" title="class in io.jenetics">AltererResult</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab4">
<div class="block">Represents the result pair of a <a href="Alterer.html#alter(io.jenetics.util.Seq,long)"><code>Alterer.alter(Seq, long)</code></a> call, which
 consists of the altered population and the number of altered individuals.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="AnyChromosome.html" title="class in io.jenetics">AnyChromosome</a>&lt;A&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block"><code>Chromosome</code> implementation, which allows to create genes without
 explicit implementing the <code>Chromosome</code> interface.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="AnyGene.html" title="class in io.jenetics">AnyGene</a>&lt;A&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block"><code>Gene</code> implementation, which allows to create genes without explicit
 implementing the <code>Gene</code> interface.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="BitChromosome.html" title="class in io.jenetics">BitChromosome</a></div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Implementation of the <i>classical</i> BitChromosome.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab3"><a href="BitGene.html" title="enum class in io.jenetics">BitGene</a></div>
<div class="col-last odd-row-color class-summary class-summary-tab3">
<div class="block">Implementation of a BitGene.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="BoltzmannSelector.html" title="class in io.jenetics">BoltzmannSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>N extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Number.html" title="class or interface in java.lang" class="external-link">Number</a> &amp; <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super N&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">
 In this <code>Selector</code>, the probability for selection is defined as.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab1"><a href="BoundedChromosome.html" title="interface in io.jenetics">BoundedChromosome</a>&lt;A extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super A&gt;,<wbr>G extends <a href="BoundedGene.html" title="interface in io.jenetics">BoundedGene</a>&lt;A,<wbr>G&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab1">
<div class="block">Chromosome interface for <code>BoundedGene</code>s.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab1"><a href="BoundedGene.html" title="interface in io.jenetics">BoundedGene</a>&lt;A extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super A&gt;,<wbr>G extends <a href="BoundedGene.html" title="interface in io.jenetics">BoundedGene</a>&lt;A,<wbr>G&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab1">
<div class="block">Base interface for genes where the alleles are bound by a minimum and a
 maximum value.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="CharacterChromosome.html" title="class in io.jenetics">CharacterChromosome</a></div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">CharacterChromosome which represents character sequences.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="CharacterGene.html" title="class in io.jenetics">CharacterGene</a></div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Character gene implementation.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab1"><a href="Chromosome.html" title="interface in io.jenetics">Chromosome</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab1">
<div class="block">A chromosome consists of one or more genes.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="CombineAlterer.html" title="class in io.jenetics">CombineAlterer</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Alters a chromosome by replacing two genes by the result of a given
 <em>combiner</em> function.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="Crossover.html" title="class in io.jenetics">Crossover</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">
 Performs a <a href="http://en.wikipedia.org/wiki/Crossover_%28genetic_algorithm%29">
 Crossover</a> of two <a href="Chromosome.html" title="interface in io.jenetics"><code>Chromosome</code></a>.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="DoubleChromosome.html" title="class in io.jenetics">DoubleChromosome</a></div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Numeric chromosome implementation which holds 64 bit floating point numbers.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="DoubleGene.html" title="class in io.jenetics">DoubleGene</a></div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">Implementation of the NumericGene which holds a 64 bit floating point number.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="EliteSelector.html" title="class in io.jenetics">EliteSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">The <code>EliteSelector</code> copies a small proportion of the fittest candidates,
 without changes, into the next generation.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="EnumGene.html" title="class in io.jenetics">EnumGene</a>&lt;A&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">
 Gene which holds enumerable (countable) genes.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="ExponentialRankSelector.html" title="class in io.jenetics">ExponentialRankSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">
 An alternative to the "weak" <code>LinearRankSelector</code> is to assign
 survival probabilities to the sorted individuals using an exponential
 function.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="GaussianMutator.html" title="class in io.jenetics">GaussianMutator</a>&lt;G extends <a href="NumericGene.html" title="interface in io.jenetics">NumericGene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">The GaussianMutator class performs the mutation of a <a href="NumericGene.html" title="interface in io.jenetics"><code>NumericGene</code></a>.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab1"><a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;A,<wbr>G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;A,<wbr>G&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab1">
<div class="block">Genes are the atoms of the <em>Jenetics</em> library.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="Genotype.html" title="class in io.jenetics">Genotype</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">The central class the GA is working with, is the <code>Genotype</code>.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="IntegerChromosome.html" title="class in io.jenetics">IntegerChromosome</a></div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">Numeric chromosome implementation which holds 32-bit integer numbers.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="IntegerGene.html" title="class in io.jenetics">IntegerGene</a></div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">NumericGene implementation which holds a 32-bit integer number.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="IntermediateCrossover.html" title="class in io.jenetics">IntermediateCrossover</a>&lt;G extends <a href="NumericGene.html" title="interface in io.jenetics">NumericGene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">This alterer takes two chromosome (treating it as vectors) and creates a
 linear combination of these vectors as result.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="LinearRankSelector.html" title="class in io.jenetics">LinearRankSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">
 In linear-ranking selection the individuals are sorted according to their
 fitness values.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="LineCrossover.html" title="class in io.jenetics">LineCrossover</a>&lt;G extends <a href="NumericGene.html" title="interface in io.jenetics">NumericGene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">This alterer takes two chromosome (treating it as vectors) and creates a
 linear combination of this vectors as result.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="LongChromosome.html" title="class in io.jenetics">LongChromosome</a></div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">Numeric chromosome implementation which holds 64-bit integer numbers.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="LongGene.html" title="class in io.jenetics">LongGene</a></div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">NumericGene implementation which holds a 64-bit integer number.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="MeanAlterer.html" title="class in io.jenetics">MeanAlterer</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt; &amp; <a href="util/Mean.html" title="interface in io.jenetics.util">Mean</a>&lt;G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">Alters a chromosome by replacing two genes by its mean value.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="MonteCarloSelector.html" title="class in io.jenetics">MonteCarloSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">The Monte Carlo selector selects the individuals from a given population
 randomly.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="MultiPointCrossover.html" title="class in io.jenetics">MultiPointCrossover</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block"><strong>Multiple point crossover</strong></div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="Mutator.html" title="class in io.jenetics">Mutator</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">This class is for mutating a chromosomes of an given population.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab4"><a href="MutatorResult.html" title="class in io.jenetics">MutatorResult</a>&lt;T&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab4">
<div class="block">Represents the result pair of one of the four <code>Mutator.mutate</code> calls.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab1"><a href="NumericChromosome.html" title="interface in io.jenetics">NumericChromosome</a>&lt;N extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Number.html" title="class or interface in java.lang" class="external-link">Number</a> &amp; <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super N&gt;,<wbr>G extends <a href="NumericGene.html" title="interface in io.jenetics">NumericGene</a>&lt;N,<wbr>G&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab1">
<div class="block">Numeric chromosome interface.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab1"><a href="NumericGene.html" title="interface in io.jenetics">NumericGene</a>&lt;N extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Number.html" title="class or interface in java.lang" class="external-link">Number</a> &amp; <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super N&gt;,<wbr>G extends <a href="NumericGene.html" title="interface in io.jenetics">NumericGene</a>&lt;N,<wbr>G&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab1">
<div class="block">Base interface for numeric genes.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab3"><a href="Optimize.html" title="enum class in io.jenetics">Optimize</a></div>
<div class="col-last even-row-color class-summary class-summary-tab3">
<div class="block">This <code>enum</code> determines whether the GA should maximize or minimize the
 fitness function.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="PartialAlterer.html" title="class in io.jenetics">PartialAlterer</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">This alterer wraps a given alterer which works on a given section of the
 genotype's chromosomes.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="PartiallyMatchedCrossover.html" title="class in io.jenetics">PartiallyMatchedCrossover</a>&lt;T,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">The <code>PartiallyMatchedCrossover</code> (PMX) guarantees that all <a href="Gene.html" title="interface in io.jenetics"><code>Gene</code></a>s
 are found exactly once in each chromosome.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="PermutationChromosome.html" title="class in io.jenetics">PermutationChromosome</a>&lt;T&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">This chromosome can be used to model permutations of a given (sub) set of
 alleles.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="Phenotype.html" title="class in io.jenetics">Phenotype</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">The <code>Phenotype</code> consists of a <a href="Genotype.html" title="class in io.jenetics"><code>Genotype</code></a>, the current generation
 and an optional fitness value.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="ProbabilitySelector.html" title="class in io.jenetics">ProbabilitySelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">Probability selectors are a variation of fitness proportional selectors and
 selects individuals from a given population based on it's selection
 probability <i>P(i)</i>.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="Recombinator.html" title="class in io.jenetics">Recombinator</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">
 An enhanced genetic algorithm (EGA) combine elements of existing solutions in
 order to create a new solution, with some of the properties of each parent.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="RouletteWheelSelector.html" title="class in io.jenetics">RouletteWheelSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>N extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Number.html" title="class or interface in java.lang" class="external-link">Number</a> &amp; <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super N&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">The roulette-wheel selector is also known as fitness proportional selector,
 but in the <em>Jenetics</em> library it is implemented as probability selector.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab1"><a href="Selector.html" title="interface in io.jenetics">Selector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab1">
<div class="block">Selectors are responsible for selecting a given number of individuals from
 the population.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="SinglePointCrossover.html" title="class in io.jenetics">SinglePointCrossover</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block"><strong>Single point crossover</strong></div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="StochasticUniversalSelector.html" title="class in io.jenetics">StochasticUniversalSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>N extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Number.html" title="class or interface in java.lang" class="external-link">Number</a> &amp; <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super N&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block"><code>StochasticUniversalSelector</code> is a method for selecting a
 population according to some given probability in a way that minimize chance
 fluctuations.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="SwapMutator.html" title="class in io.jenetics">SwapMutator</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">The <code>SwapMutation</code> changes the order of genes in a chromosome, with the
 hope of bringing related genes closer together, thereby facilitating the
 production of building blocks.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="TournamentSelector.html" title="class in io.jenetics">TournamentSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">In tournament selection the best individual from a random sample of <i>s</i>
 individuals is chosen from the population <i>P<sub>g</sub></i>.</div>
</div>
<div class="col-first odd-row-color class-summary class-summary-tab2"><a href="TruncationSelector.html" title="class in io.jenetics">TruncationSelector</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last odd-row-color class-summary class-summary-tab2">
<div class="block">In truncation selection individuals are sorted according to their fitness.</div>
</div>
<div class="col-first even-row-color class-summary class-summary-tab2"><a href="UniformCrossover.html" title="class in io.jenetics">UniformCrossover</a>&lt;G extends <a href="Gene.html" title="interface in io.jenetics">Gene</a>&lt;?,<wbr>G&gt;,<wbr>C extends <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html" title="class or interface in java.lang" class="external-link">Comparable</a>&lt;? super C&gt;&gt;</div>
<div class="col-last even-row-color class-summary class-summary-tab2">
<div class="block">The uniform crossover uses swaps single genes between two chromosomes, instead
 of whole ranges as in single- and multi-point crossover.</div>
</div>
</div>
</div>
</div>
</li>
</ul>
</section>
</main>
<footer role="contentinfo">
<hr>
<p class="legal-copy"><small>&copy; 2007-2022 Franz Wilhelmst&ouml;tter  &nbsp;<i>(2022-02-21 18:47)</i></small></p>
</footer>
</div>
</div>
</body>
</html>
